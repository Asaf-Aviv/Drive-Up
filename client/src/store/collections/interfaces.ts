import { MovieShowcase } from '../movies/interfaces';

export interface Collection {
  id: number;
  name: string;
  overview: string;
  poster_path: string | null;
  backdrop_path: string | null;
  parts: MovieShowcase[];
}
