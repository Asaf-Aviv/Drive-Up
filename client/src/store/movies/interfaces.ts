export interface Results<T> {
  results: T;
  page: number;
  total_pages: number;
  total_results: number;
}

export interface Genre {
  id: number;
  name: string;
}

export interface Series {
  id: number;
  name: string;
  poster_path: string;
  backdrop_path: string;
}

export interface Company {
  name: string;
  id: number;
  logo_path: string | null;
  origin_country: string;
}

export interface Image {
  file_path: string;
  width: number;
  height: number;
}

export interface Video {
  id: string;
  key: string;
  name: string;
  site: string;
  type: string;
}

export interface MovieShowcase {
  poster_path: string | null;
  adult: boolean;
  overview: string;
  release_date: string;
  genre_ids: number[];
  id: number;
  original_title: string;
  original_language: string;
  title: string;
  backdrop_path: string | null;
  popularity: number;
  vote_count: number;
  video: boolean;
  vote_average: number;
}

export interface Movie {
  adult: boolean;
  backdrop_path: string | null;
  belongs_to_collection: Series | null;
  budget: number;
  genres: Genre[];
  homepage: string | null;
  id: number;
  images: {
    posters: Image[];
    backdrops: Image[];
  };
  videos: {
    results: Video[];
  };
  similar: Results<MovieShowcase[]>;
  recommendations: Results<MovieShowcase[]>;
  original_language: string;
  original_title: string;
  overview: string;
  popularity: number;
  poster_path: string | null;
  production_companies: Company[];
  release_date: string;
  revenue: number;
  runtime: number | null;
  tagline: string | null;
  title: string;
  video: boolean;
  vote_average: number;
  vote_count: number;
}
