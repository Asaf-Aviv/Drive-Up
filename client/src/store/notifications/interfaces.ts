type NotificationVariant = 'error' | 'success' | 'info' | 'warning';

export interface Notification {
  id: string;
  variant: NotificationVariant;
  message: string;
}
