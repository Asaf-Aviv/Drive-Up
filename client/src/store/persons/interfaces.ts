export interface PersonSummary {
  profile_path: string;
  adult: boolean;
  id: number;
  name: string;
  popularity: number;
  known_for_department: string;
  gender: 0 | 1 | 2;
}
