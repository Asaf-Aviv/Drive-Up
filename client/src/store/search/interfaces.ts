import { MovieShowcase } from '../movies/interfaces';
import { PersonSummary } from '../persons/interfaces';
import { ShowShowcase } from '../shows/interfaces';

interface MediaType<T = 'movie' | 'tv' | 'person'> {
  media_type: T;
}

export type SearchResults =
  | MovieShowcase & MediaType<'movie'>
  | ShowShowcase & MediaType<'tv'>
  | PersonSummary & MediaType<'person'>;
