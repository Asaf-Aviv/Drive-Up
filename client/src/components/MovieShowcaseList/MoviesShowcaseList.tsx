import React, { useEffect, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { List, ListItem } from '@material-ui/core';
import useShallowEqualSelector from '../../hooks/useShallowEqualSelector';
import { requestMovies } from '../../store/movies/actions';
import { useOnVisibilityTrigger } from '../../hooks/useOnVisibilityTrigger';
import MovieShowcase from '../MovieShowcase/MovieShowcase';
import { MovieShowcase as IMovieShowcase } from '../../store/movies/interfaces';
import RetryButton from '../RetryButton';

const MovieShowcaseList: React.FC = () => {
  const {
    page, loading, movies, error,
  } = useShallowEqualSelector(state => state.movies);
  const dispatch = useDispatch();
  const fetchNextPageTrigger = useRef<HTMLElement>(null);

  const fetchNextMoviesPage = () => {
    if (movies.length) return;

    dispatch(requestMovies(page + 1));
  };

  useEffect(fetchNextMoviesPage, []);

  useOnVisibilityTrigger(fetchNextPageTrigger, fetchNextMoviesPage);

  const renderMovieShowcase = (movie: IMovieShowcase) => (
    <ListItem key={movie.id}>
      <MovieShowcase {...movie} />
    </ListItem>
  );

  return (
    <main>
      <List>
        {movies.map(renderMovieShowcase)}
        {!error && !loading && <span ref={fetchNextPageTrigger} />}
        {loading && <span>Loading...</span>}
        {error && <RetryButton onClick={fetchNextMoviesPage} />}
      </List>
    </main>
  );
};

export default MovieShowcaseList;
