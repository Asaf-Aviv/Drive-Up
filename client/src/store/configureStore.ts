import {
  createStore,
  applyMiddleware,
  Middleware,
  Reducer,
} from 'redux';
import thunkMiddleware from 'redux-thunk';
import logger from 'redux-logger';
import createSagaMiddleware from 'redux-saga';
import { composeWithDevTools } from 'redux-devtools-extension';
import rootSaga from './sagas/rootSaga';

export default function configureStore(rootReducer: Reducer, preloadedState = {}) {
  const devMode = process.env.NODE_ENV !== 'production';

  const sagaMiddleware = createSagaMiddleware();

  const middlewares: Middleware[] = [sagaMiddleware, thunkMiddleware];

  if (devMode) {
    middlewares.push(logger);
  }

  const middleWareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(
    rootReducer,
    preloadedState,
    devMode
      ? composeWithDevTools(middleWareEnhancer)
      : middleWareEnhancer,
  );

  sagaMiddleware.run(rootSaga);

  return store;
}
