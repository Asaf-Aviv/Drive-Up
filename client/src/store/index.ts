import {
  combineReducers,
  createStore,
  applyMiddleware,
  Middleware,
} from 'redux';
import thunkMiddleware from 'redux-thunk';
import logger from 'redux-logger';
import createSagaMiddleware, { SagaMiddleware } from 'redux-saga';
import { composeWithDevTools } from 'redux-devtools-extension';
import notificationsReducer from './notifications/reducers';
import moviesReducer from './movies/reducers';

const rootReducer = combineReducers({
  notifications: notificationsReducer,
  movies: moviesReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export default () => {
  const devMode = process.env.NODE_ENV !== 'production';

  const sagaMiddleware = createSagaMiddleware();

  const middlewares: Middleware[] = [sagaMiddleware, thunkMiddleware];

  if (devMode) {
    middlewares.push(logger);
  }

  const middleWareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(
    rootReducer,
    devMode
      ? composeWithDevTools(middleWareEnhancer)
      : middleWareEnhancer,
  );

  // sagaMiddleware.run()

  return store;
};
