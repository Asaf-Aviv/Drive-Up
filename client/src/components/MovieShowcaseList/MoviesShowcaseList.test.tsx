import React from 'react';
import 'intersection-observer';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { Store } from 'redux';
import { MemoryRouter } from 'react-router';
import { MovieShowcase } from '../../store/movies/interfaces';
import MoviesShowcaseList from './MoviesShowcaseList';
import configureStore from '../../store/configureStore';
import moviesReducer from '../../store/movies/reducers';
import { rootReducer } from '../../store';

const movies: MovieShowcase[] = [
  {
    adult: false,
    backdrop_path: '/4W0FnjSGn4x0mKZlBRx8OjFxQUM.jpg',
    genre_ids: [27],
    id: 474350,
    original_language: 'en',
    original_title: 'It Chapter Two',
    overview: '27 years after',
    popularity: 450.453,
    poster_path: '/zfE0R94v1E8cuKAerbskfD3VfUt.jpg',
    release_date: '2019-09-06',
    title: 'It Chapter Two',
    video: false,
    vote_average: 7.4,
    vote_count: 135,
  },
  {
    adult: false,
    backdrop_path: '/4W0FnjSGn4x0mKZlBRx8OjFxQUM.jpg',
    genre_ids: [27],
    id: 52374350,
    original_language: 'en',
    original_title: 'It Chapter Three',
    overview: '3 years after',
    popularity: 450.453,
    poster_path: '/zfE0R94v1E8cuKAerbskfD3VfUt.jpg',
    release_date: '2019-09-06',
    title: 'It Chapter Two',
    video: false,
    vote_average: 7.4,
    vote_count: 135,
  },
];

describe('MoviesShowcaseList component', () => {
  let store: Store;

  it('renders correctly', () => {
    store = configureStore(rootReducer);

    const wrapper = mount(
      <Provider store={store}>
        <MoviesShowcaseList />
      </Provider>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should render Loading... when loading state is true', () => {
    store = configureStore(moviesReducer, {
      movies: {
        page: 0,
        loading: true,
        error: false,
        movies: [],
      },
    });

    const wrapper = mount(
      <Provider store={store}>
        <MoviesShowcaseList />
      </Provider>
    );

    const spanText = wrapper.find('span').text();

    expect(spanText).toEqual('Loading...');
  });

  it('should render RetryButton component when error state is true', () => {
    store = configureStore(moviesReducer, {
      movies: {
        page: 0,
        loading: false,
        error: true,
        movies: [],
      },
    });

    const wrapper = mount(
      <Provider store={store}>
        <MoviesShowcaseList />
      </Provider>
    );

    expect(wrapper.find('RetryButton')).toHaveLength(1);
  });

  it('should render movies list', () => {
    store = configureStore(moviesReducer, {
      movies: {
        page: 0,
        loading: false,
        error: false,
        movies,
      },
    });

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <MoviesShowcaseList />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('MovieShowcase')).toHaveLength(2);
  });
});
