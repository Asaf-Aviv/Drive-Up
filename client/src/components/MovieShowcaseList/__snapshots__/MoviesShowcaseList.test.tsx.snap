// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesShowcaseList component renders correctly 1`] = `
"<Provider store={{...}}>
  <MovieShowcaseList>
    <main>
      <WithStyles(ForwardRef(List))>
        <ForwardRef(List) classes={{...}}>
          <ul className=\\"MuiList-root MuiList-padding\\">
            <span>
              Loading...
            </span>
          </ul>
        </ForwardRef(List)>
      </WithStyles(ForwardRef(List))>
    </main>
  </MovieShowcaseList>
</Provider>"
`;
